exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java

```diff
@@ -115,6 +115,8 @@
   private static final ObjectMapper MAPPER = new ObjectMapper();
   private static final long RPC_WAIT_IN_MSECS_PER_FRAGMENT = 5000;
 
+  private Thread myThreadRef; // Thread that is currently executing the Foreman.
```

Comment:
I think you need either volatile or an atomic reference here since moveToState can be executed by a thread other than the one running the fragment.